src/Database/Expression/WindowExpression.php

```diff
@@ -19,25 +19,161 @@
 use Cake\Database\ExpressionInterface;
 use Cake\Database\ValueBinder;
 use Closure;
+use InvalidArgumentException;
 
 /**
  * This represents a SQL window expression used by aggregate and window functions.
  */
-class WindowExpression implements ExpressionInterface
+class WindowExpression implements ExpressionInterface, WindowInterface
 {
+    /**
+     * @var \Cake\Database\ExpressionInterface[]
+     */
+    protected $partitions = [];
+
+    /**
+     * @var \Cake\Database\Expression\OrderByExpression|null
+     */
+    protected $order;
+
+    /**
+     * @inheritDoc
+     */
+    public function partition($partitions)
+    {
+        if (!$partitions) {
+            return $this;
+        }
+
+        if (!is_array($partitions)) {
+            $partitions = [$partitions];
+        }
+
+        foreach ($partitions as &$partition) {
+            if (is_string($partition)) {
+                $partition = new IdentifierExpression($partition);
+            }
+
+            if (
+                $partition instanceof WindowExpression ||
+                ($partition instanceof AggregateExpression && $partition->getWindow())
+            ) {
+                throw new InvalidArgumentException('Window partition cannot contain other window functions.');
```

Comment:
This should happen in sql() since the user could add the window after. Not sure if it is important to validate this or make the user responsible for valid sql.